### PR TITLE
Generate Kotlin DSL docs using Dokkatoo

### LINKED ISSUE #22953: Generated Gradle Kotlin DSL API docs might be outdated
Are there plans to update the [generated Gradle Kotlin DSL API documentation](https://gradle.github.io/kotlin-dsl-docs/api/)? I've noticed that the last commit in the gradle/kotlin-dsl-docs repository appears to be from [4 years ago](https://github.com/gradle/kotlin-dsl-docs/commit/f8ada733747c63632d652933af1812bc8221325e), and the Gradle Kotlin DSL has likely been updated since then.

----

Also see
* https://gradle.slack.com/archives/C10J6STFZ/p1673458853606139

### PR DESCRIPTION
**Pull request by jbartok**
* Fixes #22953

----
* [x] Generate dokka html in `:docs` for public API + Kotlin DSL
* [x] Include sources generated in `:kotlin-dsl`
* [x] Include sources generated at runtime by `ApiExtensionsJar.kt`
* [x] Put the generated html in the overall docs directory tree, available using `:docs:serveDocs`
* [x] Update userguide menu link to `Kotlin DSL Reference` to point to the dokka html
* [x] Make sure search works when hosted (try with `:docs:serveDocs`)
* [x] See what we can do with the left pane navigation in the dokka html output
* [x] See how we can inject CSS for future adaptation of the style to ours
* [x] Investigate showing Kotlin snippets in generated docs instead of Groovy ones - move to separate issue
* [x] Investigate having a preamble page as the entry point to the Kotlin reference (like the Groovy one)
* [x] Find a solution for Dokka not handling the `@value` javadoc tag (see Settings for an example) - move to separate issue
* [x] Upgrade to Dokkatoo 1.0.0
* [x] Validate that the Dokka task are up-to-date (second run) and from cache (after clean)

----

Preview available here
https://builds.gradle.org/repository/download/Gradle_Master_Check_BuildDistributions/64369000:id/distributions/gradle-8.2-docs.zip!/gradle-8.2-20230405153748%2B0000/docs/kotlin-dsl/index.html

### CODE REVIEW

File: build-logic/documentation/src/main/groovy/gradlebuild/docs/GradleDocumentationExtension.java [hunk 1]
```diff
@@ -32,13 +32,15 @@ public abstract class GradleDocumentationExtension {
...
     private final Javadocs javadocs;
+    private final Dokkadocs dokkadocs;
 
     @Inject
     public GradleDocumentationExtension(ObjectFactory objects) {
         releaseNotes = objects.newInstance(ReleaseNotes.class);
         userManual = objects.newInstance(UserManual.class);
         dslReference = objects.newInstance(DslReference.class);
         javadocs = objects.newInstance(Javadocs.class);
+        dokkadocs = objects.newInstance(Dokkadocs.class);
```
Comment (eskatos) [inline build-logic/documentation/src/main/groovy/gradlebuild/docs/GradleDocumentationExtension.java:43]: Let's call this `KotlinDslReference` instead

File: build-logic/documentation/src/main/groovy/gradlebuild/docs/GradleBuildDocumentationPlugin.java [hunk 2]
```diff
@@ -73,10 +75,15 @@ private void applyConventions(Project project, TaskContainer tasks, ObjectFactor
             // Javadocs reference goes into javadoc/
             task.from(extension.getJavadocs().getRenderedDocumentation(), sub -> sub.into("javadoc"));
 
+            // Dokka Kotlin DSL reference goes into dokka/
+            task.from(extension.getDokkadocs().getRenderedDocumentation(), sub -> sub.into("dokka"));
```
Comment (eskatos) [inline build-logic/documentation/src/main/groovy/gradlebuild/docs/GradleBuildDocumentationPlugin.java:79]: Let's not name the folder and thus URL path fragment `dokka` but `kotlin-dsl`  instead.

File: build-logic/documentation/src/main/groovy/gradlebuild/docs/GradleDokkaPlugin.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import dev.adamko.dokkatoo.DokkatooExtension;
+import dev.adamko.dokkatoo.dokka.parameters.DokkaSourceSetSpec;
+import dev.adamko.dokkatoo.formats.DokkatooHtmlPlugin;
+import org.gradle.api.Action;
+import org.gradle.api.NamedDomainObjectContainer;
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.jetbrains.dokka.Platform;
+
+public class GradleDokkaPlugin implements Plugin<Project> {
```
Comment (eskatos) [inline build-logic/documentation/src/main/groovy/gradlebuild/docs/GradleDokkaPlugin.java:29]: Let's name this `GradleKotlinDslReferencePlugin`  instead

File: build-logic/basics/src/main/kotlin/gradlebuild/basics/PublicApi.kt [hunk 4]
```diff
@@ -49,4 +49,6 @@ object PublicApi {
     )
 
     val excludes = listOf("**/internal/**")
+
+    val kotlinIncludes = listOf("org/gradle/kotlin/**")
```
Comment (eskatos) [inline build-logic/basics/src/main/kotlin/gradlebuild/basics/PublicApi.kt:53]: :x: This is unused now, isn't it?

File: build-logic/documentation/build.gradle.kts [hunk 5]
```diff
@@ -20,6 +20,8 @@ dependencies {
     implementation("org.asciidoctor:asciidoctor-gradle-jvm")
     implementation("org.asciidoctor:asciidoctorj")
     implementation("org.asciidoctor:asciidoctorj-pdf")
+    implementation("dev.adamko.dokkatoo:dokkatoo-plugin:0.0.4")
+    implementation("org.jetbrains.dokka:dokka-core:1.8.10")
```
Comment (eskatos) [inline build-logic/documentation/build.gradle.kts:24]: :x: These versions should be moved to the build-logic platform

**Review by eskatos — CHANGES_REQUESTED** [review]

File: build-logic/basics/src/main/kotlin/gradlebuild/basics/PublicApi.kt (hunk 4 above)
Comment (jbartok) [reply build-logic/basics/src/main/kotlin/gradlebuild/basics/PublicApi.kt:53]: Yes, since you removed its usage. Cleaning it up.

**Comment by jbartok** [comment]
@eskatos One idea I could come up with for having Kotlin samples in all Kotlin DSL docs, even when the documentation is coming from Java classes and their Javadoc is to use the [@snippet tag for JavaDoc](https://openjdk.org/jeps/413). 

We would need to rewrite all Javadoc code snippets in the Gradle API to use the `@snippet` Javadoc tag. The snippets would be **external** and with **regions**, pretty much what we already do for user manual code samples. This would make them testable too, which is an added benefit. 

Since the content would be in files, we could have one set of snippet files to use when building Javadoc and another one when building KDoc/Dokka.

The snippet feature was introduced into Javadoc in Java 18, so we would need to use that, but that should be ok to do via toolchains, I think.

File: build-logic/documentation/src/main/groovy/gradlebuild/docs/GradleKotlinDslReferencePlugin.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+        NamedDomainObjectContainer<DokkaSourceSetSpec> sourceSets = getDokkatooExtension(project).getDokkatooSourceSets();
+        sourceSets.register("kotlin_dsl", spec -> {
+            spec.getDisplayName().set("Gradle Kotlin DSL");
+            spec.getSourceRoots().from(extension.getKotlinDslSource());
+            spec.getSourceRoots().from(runtimeExtensions.flatMap(GradleKotlinDslRuntimeGeneratedSources::getGeneratedSources));
+            spec.getClasspath().from(extension.getClasspath());
+            spec.getClasspath().from(runtimeExtensions.flatMap(GradleKotlinDslRuntimeGeneratedSources::getGeneratedClasses));
+            spec.getAnalysisPlatform().set(Platform.jvm);
+            spec.getIncludes().from(extension.getSourceRoot().file("kotlin/Module.md").get().getAsFile().getAbsolutePath());
```
Comment (eskatos) [inline build-logic/documentation/src/main/groovy/gradlebuild/docs/GradleKotlinDslReferencePlugin.java:80]: Couldn't this be made more lazy?

File: subprojects/docs/src/docs/userguide/api/kotlin_dsl.adoc [hunk 7]
```diff
@@ -175,13 +175,16 @@ This is so that the IDE knows what type of object "backs" the script, be it link
...
+** `org.gradle.kotlin.dsl`
+** `org.gradle.kotlin.dsl.plugins.dsl`
+** `org.gradle.kotlin.dsl.precompile`
 
 [CAUTION]
 .Avoid using internal Kotlin DSL APIs
 ====
 Use of internal Kotlin DSL APIs in plugins and build scripts has the potential to break builds when either Gradle or plugins change.
-The link:{kotlinDslPath}/[Kotlin DSL API] extends the <<authoring_maintainable_build_scripts#sec:avoiding_gradle_internal_apis,Gradle public API>> with the types listed in the https://gradle.github.io/kotlin-dsl-docs/api/[corresponding API docs] that are in the `org.gradle.kotlin.dsl` or `org.gradle.kotlin.dsl.plugins.dsl` packages (but not subpackages of those).
+The link:{kotlinDslPath}/[Kotlin DSL API] extends the <<authoring_maintainable_build_scripts#sec:avoiding_gradle_internal_apis,Gradle public API>> with the types listed in the https://gradle.github.io/kotlin-dsl-docs/api/[corresponding API docs] that are in the packages listed above (but not subpackages of those).
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/api/kotlin_dsl.adoc:187]: 👍

**Comment by eskatos** [comment]
@bot-gradle test BD

**Comment by eskatos** [comment]
Preview available here
https://builds.gradle.org/repository/download/Gradle_Master_Check_BuildDistributions/64369000:id/distributions/gradle-8.2-docs.zip!/gradle-8.2-20230405153748%2B0000/docs/kotlin-dsl/index.html

**Review by eskatos — APPROVED** [review]
This is a great milestone!

**Comment by hfhbd** [comment]
Any plans to add the link to the source code? See https://github.com/adamko-dev/dokkatoo/blob/693eeaa5900f77acb1fe7e9376d8fb3463fdd597/examples/gradle-example/dokkatoo/build.gradle.kts#L15-L23 for the configs or https://csv.softwork.app/kotlinx-serialization-csv/app.softwork.serialization.csv/-c-s-v-format/decode-from-string.html for a sample, the `(source)` link.

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Pull request merged**